packages/loopring_v3/circuit/Gadgets/MathGadgets.h

```diff
@@ -314,6 +461,46 @@ class AddGadget : public GadgetT
     }
 };
 
+// A - B = sub with A, B and sub >= 0
+class SubGadget : public GadgetT
+{
+public:
+    UnsafeSubGadget unsafeSub;
+    libsnark::dual_variable_gadget<FieldT> rangeCheck;
+
+    SubGadget(
+        ProtoboardT& pb,
+        const VariableT& A,
+        const VariableT& B,
+        unsigned int n,
+        const std::string& prefix
+    ) :
+        GadgetT(pb, prefix),
+
+        unsafeSub(pb, A, B, FMT(prefix, ".unsafeAdd")),
+        rangeCheck(pb, unsafeSub.result(), n, FMT(prefix, ".rangeCheck"))
```

Comment:
Is it correct to say this rangeCheck basically means:
`require(unsafeSub.result() <= 2^n -1);`?

If so, I think our AddGadgent/SubGadgent is simpler than the implementation in subadd_gadget, unless we missed something.

Reply:
![Screen Shot 2020-07-28 at 22 32 06](https://user-images.githubusercontent.com/1710087/88679613-2cdd1100-d122-11ea-97ed-37bc11029d35.jpg)

Maybe we can also checkout the ZCash range check logic.

Reply:
![Screen Shot 2020-07-28 at 22 37 36](https://user-images.githubusercontent.com/1710087/88680316-f0f67b80-d122-11ea-9a71-799f63280ae2.jpg)

Reply:
> Is it correct to say this rangeCheck basically means:
require(unsafeSub.result() <= 2^n -1);?

Yes.

> If so, I think our AddGadgent/SubGadgent is simpler than the implementation in subadd_gadget, unless we missed something.

Yes I think so.

> Maybe we can also checkout the ZCash range check logic.

By using the libsnark gadgets we're kind of automatically doing what they are doing regardless. And we mostly/only care only about powers of 2 so I think we're doing things the best we can.

Reply:
> A.3.2.2

This is for range checks against a constant value which is not a power of 2, which is more expensive than a range check against a constant value which is a power of 2. Would rarely be usable if ever, so did not implement this.